sls-versions/src/main/java/com/palantir/sls/versions/Parsers.java

```diff
@@ -0,0 +1,124 @@
+ *
+ * We bit-pack these two integer values into a single long using {@link Parsers#ok} and {@link Parsers#fail} functions
+ * because primitive longs live on the stack and don't impact GC.
+ */
+final class Parsers {
+
+    static final int MAGIC_X_NUMBER = -1;
+    private static final long INT_MASK = (1L << 32) - 1;
+
+    private Parsers() {}
+
+    // "x" is signified by the magic negative number -1, which is distinct from Integer.MIN_VALUE which is a failure
+    static long numberOrX(String string, int startIndex) {
+        long xResult = literalX(string, startIndex);
+        if (isOk(xResult)) {
+            return ok(getIndex(xResult), MAGIC_X_NUMBER);
+        }
+
+        long numberResult = number(string, startIndex);
+        if (isOk(numberResult)) {
+            return numberResult;
+        }
+
+        return fail(startIndex);
+    }
+
+    static long number(String string, int startIndex) {
+        int next = startIndex;
+        int len = string.length();
+        while (next < len) {
+            int codepoint = string.codePointAt(next);
+            if (Character.isDigit(codepoint)) {
+                next += 1;
+            } else {
+                break;
+            }
+        }
+        if (next == startIndex) {
+            return fail(startIndex);
+        } else if (next == startIndex + 1) {
+            return ok(next, Character.digit(string.codePointAt(startIndex), 10));
+        } else {
+            try {
+                return ok(next, Integer.parseUnsignedInt(string.substring(startIndex, next)));
+            } catch (NumberFormatException e) {
+                if (e.getMessage() != null && e.getMessage().endsWith("exceeds range of unsigned int.")) {
+                    return fail(startIndex);
+                } else {
+                    throw e;
+                }
+            }
+        }
+    }
+
+    // 0 signifies success
+    private static long literalX(String string, int startIndex) {
+        if (startIndex < string.length() && string.codePointAt(startIndex) == 'x') {
+            return ok(startIndex + 1, 0);
+        } else {
+            return fail(startIndex);
+        }
+    }
+
+    static long literalDot(String string, int startIndex) {
+        if (startIndex < string.length() && string.codePointAt(startIndex) == '.') {
+            return ok(startIndex + 1, 0);
+        } else {
+            return fail(startIndex);
+        }
+    }
+
+    /**
+     * We are bit-packing two integers into a single long.  The 'index' occupies half of the bits and the 'result'
+     * occupies the other half.
+     */
+    static long ok(int index, int result) {
+        return ((long) index) << 32 | (result & INT_MASK);
+    }
+
+    static long fail(int index) {
+        return ((long) index) << 32 | (Integer.MIN_VALUE & INT_MASK);
+    }
+
+    static boolean isOk(long state) {
+        return getResult(state) != Integer.MIN_VALUE;
```

Comment:
I know this is copied code, but you might as well make a `private static final int FAILED = Integer.MIN_VALUE` to make the magic number clear.